### ISSUE TITLE
Adopting a Release Workflow and SemVer Tags

### ISSUE THREAD

**Issue by tajmone** [AnssiR66/AlanStdLib#40]
Ciao Annsi,

I wanted to bring up again the question of library version. Since the first announcement of the StdLib 2.1 release on Yahoo groups, up to the present day, there have been quite a few changes to the library.

### Potential Problems

Although none of these changes break backward compatiblity, I think it would be wise to start following a strict versioning and release scheme (like SemVer). If not, the risk is that of having different versions of "v2.1" in the wild.

Also, for projects like my translation (which strongly depend on the StdLib code base) it's quite hard to pinpoint a specific version of the codebase, as I have to rely on single commits as a reference points. 

Of course, any commits info is lost when the code is separated from the repo's `.git` folder, so library sources reporting a same version number could actually belong to different commits (and mixing in a project source files from different commits could lead to unexpected results).

### Release Workflow Proposal

What I propose is to adopt some simple workflow (nothing like Git flow and similars), using a trunk-based approach and no long standing branches:

- use `master` to host the code base
- work on features in branches, and sqash or rebase them into master as soon as ready.
- use SemVer tags to mark release points.

Optionally, we could have a `releases` branch where at each release we squash `master` into it, in order to have a dedicated branch with a clean and linear history where each commit is a release point (allowing for bug fixes if needed).

This way, every time a new release is ready (wether MAJOR, MINOR or PATCH) we can use GitHub's Web UI to create a new release and tag, and GitHub will automatically add it to the releases page and create a downloadable ZIP for the release, etc..

### Benefits

Users would be able to refer to precise StdLib releases/verssions (e.g. when working with others, they could provide a tag-specific download link) and be sure that the code hansn't changed and that everyone is working with the same code base.

Of course, anyone wishing to use the bleeding-edge version of the StdLib only needs to pull `master` branch and he/she will always get access to the latest code. Since only tested features are merged into `master`, it should always be safe to use it in production.

When using a strict version and release scheme, GitHub also offers lots of cool automation tools that can be triggered when creating a new release. These include scripts that can carry out some chores, and control over what will go in the downloadable ZIP archive of the release.

#### Automation

In the long term, we might eventually consider using Continuos Integration and create some scripts to automate tests and other chores. Ultimately, we might create a "release" script that takes cares of updating all that needs to be done before every public release (update/convert documentation, make sure all files refer to the new version number, etc.). From my experience, these type of things are very useful, especially when one hasn't worked on the project for a while and then needs to update something (where the risk is that of forgetting some update chores because a long time passed since it was last done).

#### Dashboard and Planning

In terms of developement work, it would also allow us to maintain a well structured Dashboard, and keep track of what's waiting ahead of us, for we could set some Milestones for the upcoming releases by creating issues and task lists that need to be completed before the next release is ready, etc.

Especially with semantic versioning, the Dashboard and milestones are precious tools in terms of tracking MAJOR, MINOR and PATCH changes, and which features require bumping up these, and when the next release of each type is due along the line.

As a simple example, we could ask ourselves "What changes would justify releasing StdLib 2.2?" and then create a Milestone for "2.2" and start adding issues for the various pending tasks.

[SemVer]: https://semver.org/

**Comment by AnssiR66** [comment]
Yes, I don't have anything against your suggestion. For now, v2.1 is still practically unpublished (except for the Yahoo group version) but I agree it would be a good idea to start to number the versions more systematically. We can use the automatic method you mentioned.

**Comment by tajmone** [comment]
Ok, I've created a Milestone for the upcoming v2.2 (I though that to avoid confusion it might be best to start with the next MINOR version):

https://github.com/AnssiR66/AlanStdLib/milestone/2

Now we can queue to this milestones all the bug fixes, enhancements and whatever makes sense to be dealt with before creating a tagged release. So I'd start to attach to the milestone the bugs with the floor/ground and the "put in" verb on liquids.

**Comment by AnssiR66** [comment]
Hi Tristano,
OK, sounds good otherwise, but I think might be too early to talk about a v2.2 as of now. As the official version at the Alan website currently is 2.0, wouldn't it confuse users if there suddenly is a 2.2?  Anyway, any foreseeable changes to the library at this point are mainly cosmetic and bug-fixes, and there wouldn't be any big new features going to be added to the library itself, as it is rather extensive as it is presently. Any new feature(s) would probably be added through separate extensions; I was talking about this with Thomas some years ago (around when 2.0 came about). So, the way 2.1 is now at the Yahoo group, and the changes we have made since then, would maybe warrant something like 2.1x (I mean, 2.11, 2.12 or something along those lines). It's of course not a big thing in itself, I was thinking mainly  from user point-of-view.  I think that when we have got the current issues figured out, wouldn't it be logical to have something starting with 2.1x appear at the website?
-Anssi

________________________________

**Comment by tajmone** [comment]
Yes, v2.1.1 would be fine indeed, as long as backward compatibility is not broken by any of these changes (in which case MINOR version should bump up). I'd advise to move up to at least 2.1.1 to avoid the risk of different versions 2.1 having been already shared in the meantime.

Of course, 2.1.1 would still be part of the v2.1 version, as the third digit only indicates small changes and fixes.

Then, of course much depends also on wether or not you wish to use SemVer (which is strict about which digits in a version should change depending on the nature of changes), or if you rather adopt a custom versioning system (after all, it's not an API, it's an IF library), in which case anything goes as long as it's intuitive to end users.

I've updated the Milestone accordingly!
